src/injected-js/gmail/sync-compose-processor-20220909.ts

```diff
@@ -47,221 +30,288 @@ export function replaceBodyContentInComposeSendRequestBody_2022_09_09(
+}
 
-    const subject = msg[7];
-    const body = msg[8][1][0][1];
-    const actions = msg[10];
+/**
+ * Parses response body when compose window either saved a draft for the first time,
+ * updated the draft, or sent the draft.
+ * NOTE: response could contain multiple threads and messages within it
+ * even not related to a message/thread in the request body.
+ * So the calling code should find needed message manually.
+ */
+function parseCreateUpdateSendDraftResponseBody(response: any[]) {
+  const updateList = response[1]?.[5];
 
-    const hasRequiredActions =
-      intersection(actions, DRAFT_SAVING_ACTIONS).length ===
-      DRAFT_SAVING_ACTIONS.length;
+  if (!Array.isArray(updateList)) {
+    // exit cuz cannot parse
+    return [];
+  }
 
-    if (!hasRequiredActions) {
-      // exit if doesn't have required actions
-      return null;
-    }
+  return updateList
+    .map(parseResponseThread)
+    .filter(isNotNil)
+    .flatMap((parsedThread) => {
+      const { threadId, oldThreadId, parsedMessages } = parsedThread;
+
+      return parsedMessages.map((parsedMessage) => {
+        const { messageId, to, cc, bcc, actions, rfcID, oldMessageId } =
+          parsedMessage;
+
+        const actionType = actionsToComposeRequestType(actions);
+
+        if (!actionType) {
+          // unsupported actions within a message
+          return null;
+        }
+
+        return {
+          threadId,
+          messageId,
+          to,
+          cc,
+          bcc,
+          actions,
+          rfcID,
+          oldMessageId,
+          oldThreadId,
+          type: actionType,
+        };
+      });
+    })
+    .filter(isNotNil);
+}
 
-    return {
-      threadId,
-      messageId,
-      subject,
-      body,
-      actions,
-      type: 'FIRST_DRAFT_SAVE' as ComposeRequestType,
-    };
+function replaceBodyContentInSendRequestBody(
+  request: Array<any>,
+  newBodyHtmlContent: string
+) {
+  // since draftID is not passed from outside,
+  // use parse method to find a message which body needs to be replaced
+  const parsed = parseCreateUpdateSendDraftRequestBody(request);
+  if (!parsed) {
+    return null;
   }
-}
 
-/**
- * Parses request when compose window updates draft (consecutive updates after creation)
- */
-function parseUpdateDraftRequestBody(request: Array<any>) {
-  const thread =
-    request[1] && request[1][0] && request[1][0][0] && request[1][0][0][1];
+  const replaceBodyInThisMessageId = parsed.messageId;
 
-  if (thread) {
-    const threadId = parseThreadId(thread[0]);
-    if (!threadId) {
-      // exit cuz cannot parse
-      return null;
-    }
+  const updateList = request[1]?.[0];
 
-    const msg = thread[1] && thread[1][13] && thread[1][13][0];
+  if (!Array.isArray(updateList)) {
+    // exit cuz cannot parse
+    return null;
+  }
 
-    if (!Array.isArray(msg)) {
+  for (const threadWrapper of updateList) {
+    if (!Array.isArray(threadWrapper) || !Array.isArray(threadWrapper[1])) {
```

Comment:
If threadWrapper is null, this will throw, won't it? I think the older code was an optional chain.

```suggestion
    if (!threadWrapper?.[1])) {
```

Reply:
it wont throw, will keep as is